sequence_metrics/metrics.py

```diff
@@ -291,44 +341,78 @@ def strip_whitespace(y):
         "text": label_text.strip(),
         "start": new_start,
         "end": new_start + len(stripped),
-        "label": y["label"],
+        **{k: v for k, v in y.items() if k not in ["text", "start", "end"]},
     }
 
 
 def _norm_text(text: str) -> str:
     return re.sub(r"[^A-Za-z0-9]", "", text).lower()
 
 
-def fuzzy_compare(x: dict, y: dict) -> bool:
+def fuzzy_compare(x: LabeledSpan, y: LabeledSpan) -> bool:
     return _norm_text(x["text"]) == _norm_text(y["text"])
 
 
-def sequence_labeling_token_precision(true, predicted):
+def sequence_labeling_token_precision(
+    true: t.Sequence[t.Sequence[LabeledSpan]],
+    predicted: t.Sequence[t.Sequence[LabeledSpan]],
+) -> dict[str, float | None]:
     """
     Token level precision
     """
     return seq_precision(true, predicted, span_type="token")
 
 
-def sequence_labeling_token_recall(true, predicted):
+def sequence_labeling_token_recall(
+    true: t.Sequence[t.Sequence[LabeledSpan]],
+    predicted: t.Sequence[t.Sequence[LabeledSpan]],
+) -> dict[str, float | None]:
     """
     Token level recall
     """
     return seq_recall(true, predicted, span_type="token")
 
 
-def sequence_labeling_micro_token_f1(true, predicted):
+def sequence_labeling_micro_token_f1(
+    true: t.Sequence[t.Sequence[LabeledSpan]],
+    predicted: t.Sequence[t.Sequence[LabeledSpan]],
+) -> float | None:
     """
     Token level F1
     """
     return micro_f1(true, predicted, span_type="token")
 
 
-def sequences_overlap(x: dict, y: dict) -> bool:
+def check_doc_id(
+    metric_fn: Callable[[LabeledSpan, LabeledSpan], bool]
+) -> Callable[[LabeledSpan, LabeledSpan], bool]:
+    """
+    A decorator that adds doc_id check to the comparison function.
+    """
+
+    @functools.wraps(metric_fn)
+    def wrapper(true: LabeledSpan, predicted: LabeledSpan):
+        true_doc_id = true.get("doc_id", None)
+        predicted_doc_id = predicted.get("doc_id", None)
+        if true_doc_id is not None or predicted_doc_id is not None:
+            if true_doc_id is None:
+                raise ValueError("Label value is missing doc_id")
+            if predicted_doc_id is None:
+                raise ValueError("Prediction value is missing doc_id")
+            if true_doc_id != predicted_doc_id:
+                return False
+        return metric_fn(true, predicted)
+
+    return wrapper
+
+
+@check_doc_id
```

Comment:
This makes doc_id fully required -- which is desirable in this case because it then flags any cases where it isn't being appropriately wired through by causing hard errors?

Reply:
It needs to either be fully missing or fully populated. A mix of has a doc-id and not has a doc-id will raise errors.